fendermint/vm/topdown/src/finality/fetch.rs

```diff
@@ -71,83 +72,62 @@ impl<T: ParentQueryProxy + Send + Sync + 'static> ParentViewProvider for CachedF
         &self,
         from: BlockHeight,
         to: BlockHeight,
+        block_hash: &BlockHash,
     ) -> anyhow::Result<Vec<StakingChangeRequest>> {
         let mut v = vec![];
-        for h in from..=to {
+        for h in from..to {
             let mut r = self.validator_changes(h).await?;
             tracing::debug!(
                 number_of_messages = r.len(),
                 height = h,
-                "obtained validator change set",
+                "fetched validator change set",
             );
             v.append(&mut r);
         }
 
-        Ok(v)
-    }
-
-    async fn validator_changes(
-        &self,
-        height: BlockHeight,
-    ) -> anyhow::Result<Vec<StakingChangeRequest>> {
-        let r = self.inner.validator_changes(height).await?;
-
-        if let Some(v) = r {
-            return Ok(v);
-        }
-
-        let r = retry!(
-            self.config.exponential_back_off,
-            self.config.exponential_retry_limit,
-            self.parent_client
-                .get_validator_changes(height)
-                .await
-                .map(|r| r.value)
+        // now we query the validator changes at height `to` and we also checks the block hash,
+        // so that we know the parent node is reliable enough.
+        let mut r = self.validator_changes_with_check(to, block_hash).await?;
+        tracing::debug!(
+            number_of_messages = r.len(),
+            height = to,
+            "fetched validator change set",
         );
+        v.append(&mut r);
 
-        handle_null_round(r, Vec::new)
-    }
-
-    /// Should always return the top down messages, only when ipc parent_client is down after exponential
-    /// retries
-    async fn top_down_msgs(
-        &self,
-        height: BlockHeight,
-        block_hash: &BlockHash,
-    ) -> anyhow::Result<Vec<CrossMsg>> {
-        let r = self.inner.top_down_msgs(height).await?;
-
-        if let Some(v) = r {
-            return Ok(v);
-        }
-
-        let r = retry!(
-            self.config.exponential_back_off,
-            self.config.exponential_retry_limit,
-            self.parent_client
-                .get_top_down_msgs_with_hash(height, block_hash)
-                .await
-        );
-
-        handle_null_round(r, Vec::new)
+        Ok(v)
     }
 
+    /// Get top down message in the range `from` to `to`, both inclusive. It also validates the block
```

Comment:
Update here to get the same range we are having in validator changes from `prev_finality` to `latest_finality -1`